EE-2095 - Improved JSON schemas

M-casado opened the pull request:
[**EE-2095**](https://www.ebi.ac.uk/panda/jira/browse/EE-2095) ticket. Improvements done to the common classes and Array's Assay and Experiment schemas. 

For the sake of clarity, take a look at each commit's description, which will be easier to interpret.

malloryfreeberg commented on schemas/EGA.ArrayAssay.json:
```diff
@@ -78,10 +81,11 @@
       },
 
-      "ArrayAssay_relationships": {
+      "Array_assay_relationships": {
```
```suggestion
      "array_assay_relationships": {
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-invalid-1_integer-string.json:
```diff
@@ -28,34 +28,34 @@
         }
     ],
-    "ArrayAssay_relationships": [
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-invalid-2_wrong-checksum.json:
```diff
@@ -28,34 +28,34 @@
         }
     ],
-    "ArrayAssay_relationships": [
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-invalid-3_wrong-egan.json:
```diff
@@ -28,34 +28,34 @@
         }
     ],
-    "ArrayAssay_relationships": [
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-invalid-4_wrong-checksum.json:
```diff
@@ -28,34 +28,34 @@
         }
     ],
-    "ArrayAssay_relationships": [
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-invalid-5_wrong-date.json:
```diff
@@ -28,34 +28,34 @@
         }
     ],
-    "ArrayAssay_relationships": [
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-valid-1.json:
```diff
@@ -28,34 +28,34 @@
         }
     ],
-    "ArrayAssay_relationships": [
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-1_wrong-egae.json:
```diff
@@ -48,7 +48,7 @@
         }
     ],
-    "ArrayExperiment_attributes": [
+    "Array_experiment_attributes": [
```
```suggestion
    "array_experiment_attributes": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-1_wrong-egae.json:
```diff
@@ -59,32 +59,32 @@
         }
         ],
-    "ArrayExperiment_relationships": [
+    "Array_experiment_relationships": [
```
```suggestion
    "array_experiment_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-1_wrong-egae.json:
```diff
@@ -59,32 +59,32 @@
         }
     ],
-    "ArrayExperiment_protocols": [
+    "Array_experiment_protocols": [
```
```suggestion
    "array_experiment_protocols": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-2_wrong-egas.json:
```diff
@@ -47,7 +47,7 @@
         }
     ],
-    "ArrayExperiment_attributes": [
+    "Array_experiment_attributes": [
```
```suggestion
    "array_experiment_attributes": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-2_wrong-egas.json:
```diff
@@ -58,28 +58,28 @@
         }
         ],
-    "ArrayExperiment_relationships": [
+    "Array_experiment_relationships": [
```
```suggestion
    "array_experiment_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-2_wrong-egas.json:
```diff
@@ -58,28 +58,28 @@
         }
     ],
-    "ArrayExperiment_protocols": [
+    "Array_experiment_protocols": [
```
```suggestion
    "array_experiment_protocols": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-3_wrong-checksum.json:
```diff
@@ -47,7 +47,7 @@
         }
     ],
-    "ArrayExperiment_attributes": [
+    "Array_experiment_attributes": [
```
```suggestion
    "array_experiment_attributes": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-invalid-6_wrong-archive.json:
```diff
@@ -0,0 +1,111 @@
+            }
+        }
+    ],
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayAssay-valid-2.json:
```diff
@@ -0,0 +1,111 @@
+            }
+        }
+    ],
+    "Array_assay_relationships": [
```
```suggestion
    "array_assay_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-3_wrong-checksum.json:
```diff
@@ -58,32 +58,32 @@
         }
         ],
-    "ArrayExperiment_relationships": [
+    "Array_experiment_relationships": [
```
```suggestion
    "array_experiment_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-3_wrong-checksum.json:
```diff
@@ -58,32 +58,32 @@
         }
     ],
-    "ArrayExperiment_protocols": [
+    "Array_experiment_protocols": [
```
```suggestion
    "array_experiment_protocols": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-4-wrong-filename.json:
```diff
@@ -0,0 +1,90 @@
+            "encrypted_checksum": "c6779ec2960296ed9a04f08d67f64423"
+        }
+    ],
+    "Array_experiment_attributes": [
```
```suggestion
    "array_experiment_attributes": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-4-wrong-filename.json:
```diff
@@ -0,0 +1,90 @@
+        "assayed_molecule_curie": "EFO:0001456"
+    },
+    "sample_number": 12,
+    "ADF_files": [
```
```suggestion
    "adf_files": [
```

malloryfreeberg replied:
I suggest having a rule that all attributes are lower_camel_case, including acronyms. For example, you use `id` and not `ID` .

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-4-wrong-filename.json:
```diff
@@ -0,0 +1,90 @@
+            "units": "GBP"
+        }
+        ],
+    "Array_experiment_relationships": [
```
```suggestion
    "array_experiment_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-invalid-4-wrong-filename.json:
```diff
@@ -0,0 +1,90 @@
+            "r_target_type": "ArrayExperiment"
+        }
+    ],
+    "Array_experiment_protocols": [
```
```suggestion
    "array_experiment_protocols": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-valid-1.json:
```diff
@@ -47,7 +47,7 @@
         }
     ],
-    "ArrayExperiment_attributes": [
+    "Array_experiment_attributes": [
```
```suggestion
    "array_experiment_attributes": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-valid-1.json:
```diff
@@ -58,32 +58,32 @@
         }
         ],
-    "ArrayExperiment_relationships": [
+    "Array_experiment_relationships": [
```
```suggestion
    "array_experiment_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-valid-1.json:
```diff
@@ -58,32 +58,32 @@
         }
     ],
-    "ArrayExperiment_protocols": [
+    "Array_experiment_protocols": [
```
```suggestion
    "array_experiment_protocols": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-valid-2_missing-protocol-curie.json:
```diff
@@ -47,7 +47,7 @@
         }
     ],
-    "ArrayExperiment_attributes": [
+    "Array_experiment_attributes": [
```
```suggestion
    "array_experiment_attributes": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-valid-2_missing-protocol-curie.json:
```diff
@@ -58,28 +58,28 @@
         }
         ],
-    "ArrayExperiment_relationships": [
+    "Array_experiment_relationships": [
```
```suggestion
    "array_experiment_relationships": [
```

malloryfreeberg commented on schemas/validation_tests/ArrayExperiment-valid-2_missing-protocol-curie.json:
```diff
@@ -58,28 +58,28 @@
         }
     ],
-    "ArrayExperiment_protocols": [
+    "Array_experiment_protocols": [
```
```suggestion
    "array_experiment_protocols": [
```

malloryfreeberg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -27,12 +27,14 @@
         "title": "Title of the ArrayExperiment",
-        "description": "An informative experiment title that should serve as an overview of the experiment, including: used technology, samples, purpose... (e.g. Affymetrix-X microarray of human breast cancer cell line MCF-7 treated with tamoxifen compared with untreated controls). This short text can be used to call out ArrayExperiment records in searches or in displays. This element is technica ...
+        "description": "An informative experiment title that should serve as an overview of the experiment, including: used technology, samples, purpose... (e.g. 'Affymetrix-X microarray of human breast cancer cell line MCF-7 treated with tamoxifen compared with untreated controls'). This short text can be used to call out ArrayExperiment records in searches or in displays. This element is techni ...
+        "example": "Affymetrix-X microarray of human breast cancer cell line MCF-7 treated with tamoxifen compared with untreated controls"
```
I like the addition of the examples!

malloryfreeberg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -116,52 +128,61 @@
       },
 
-      "ArrayExperiment_relationships": {
+      "Array_experiment_relationships": {
```
```suggestion
      "array_experiment_relationships": {
```

malloryfreeberg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -116,52 +128,61 @@
         "title": "Number of samples of the ArrayExperiment",
-        "description": "Number of samples included in the experiment. One sample corresponds to one biological replicate [EFO:0002091] (it could be the genetic content from a single cell, a tissue… from a single individual or from several individuals). Shall not be mistaken for technical replicates [CHEBI:24432] being used several times (see https://www.ebi.ac.uk/seqdb/confluence/pages/viewpage.a ...
+        "description": "Number of samples included in the experiment. One sample corresponds to one biological replicate [EFO:0002091] (it could be the genetic content from a single cell, a tissue… from a single individual or from several individuals). Shall not be mistaken for technical replicates [CHEBI:24432] being used several times (see https://www.ebi.ac.uk/seqdb/confluence/pages/viewpage.a ...
+        "example": "30"
```
```suggestion
        "example": 30
```

malloryfreeberg replied:
If this is type `integer`, then the example shouldn't be a string

M-casado replied:
Technically true, albeit it will be interpreted as a string, since it's a keyword that we just made up for the future markdown.

malloryfreeberg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -116,52 +128,61 @@
       },
 
-      "ArrayExperiment_attributes": {
+      "Array_experiment_attributes": {
```
```suggestion
      "array_experiment_attributes": {
```

malloryfreeberg commented on schemas/EGA.ArrayExperiment.json:
```diff
@@ -116,52 +128,61 @@
       },
 
-      "ArrayExperiment_protocols": {
+      "Array_experiment_protocols": {
```
```suggestion
      "array_experiment_protocols": {
```

malloryfreeberg commented on schemas/EGA.common-definitions.json:
```diff
@@ -15,24 +15,37 @@
+            "type": "array",
+            "title": "External accessions array",
+            "description": "Custom attributes of an ArrayExperiment: reusable attributes to encode tag-value pairs (e.g. Tag being 'Targeted loci' and its Value '5:63256183-63258334') with optional units (e.g. 'base pairs'). Its properties are inherited from the common-definitions.json schema. #! Using an empty item list but defining the custom attributes as possible additionalItems we create the ...
+            "minItems": 1,
```
Does `"minItems": 1,` mean this attribute is required? Is that too strict? i.e. will there be submissions that won't have external accessions?

M-casado replied:
It does not, what it states is that, if given an ``eternal_accessions`` array, it needs to contain at least 1 item in the array. 
The requirements for the object's core are one of the following combinations:
````
"anyOf": [
          { "required": ["alias", "center_name"]},
          { "required": ["ega_accession"]},
          { "required": ["external_accessions"]}
        ]
````

malloryfreeberg reviewed (changes requested):
This was a lot of work- great job! A few comments:

- [ ] Suggested making all attribute names lower_camel_case. I really only. noticed the ones that were updated in this PR - might be good to do another full check that there are no uppercases in attribute names.
- [ ] Are there instructions for how I can check whether the "valid" example JSON documents are valid and the "invalid" ones invalid?
- [ ] I found one instance of an integer attribute having a string as the example, and suggested a change here. I suggest making sure all integer attributes indeed have integer examples. This is a check that could eventually be added to a linter.

M-casado commented:
@malloryfreeberg:

> * [x]  Suggested making all attribute names lower_camel_case. I really only. noticed the ones that were updated in this PR - might be good to do another full check that there are no uppercases in attribute names.

I scanned the three schemas and all examples with regular expressions and found no other occurrence besides the ``ADF_...`` I already fixed in my previous commit. 

> * [x]  Are there instructions for how I can check whether the "valid" example JSON documents are valid and the "invalid" ones invalid?

The instructions are within the Jupyter notebook. But since these tests are not scalable and at the beginning were intended for just development, they are expected to be deprecated in favour of the test module from [EE-2217](https://www.ebi.ac.uk/panda/jira/browse/EE-2217).

> * [x]  I found one instance of an integer attribute having a string as the example, and suggested a change here. I suggest making sure all integer attributes indeed have integer examples. This is a check that could eventually be added to a linter.

Fixed another occurrence of an example.

M-casado commented:
@ainacrg @colthom @sdelatorrep 

While we develop a tool to turn JSON into markdown (https://www.ebi.ac.uk/panda/jira/browse/EE-2212), to ease the review of this current PR, the actions taken are summarised here:

- [ ] Harmonized from mixed camelCase/snake_case to snake_case. 
- [ ] Added filetype-filename checks (based on regex and filename suffixes). 
- [ ] Modified the Term-CURIE fields so that only one is recorded and we maintain **within the schemas** a term-curie pair list for each of these terms. This will be parsed somehow in the future to create abovementioned markdown documents. 
- [ ] Added "example" keyword to use in the future for documentation purposes. 
- [ ] Added the external accession node to the object's core identifier node. 
- [ ] Added the schemas to the CHANGELOG of the repo.

malloryfreeberg reviewed (approved):
LGTM. Please merge for now as these changes are mostly just to harmonise the format of the existing schemas, and there aren't any contentious issues.
